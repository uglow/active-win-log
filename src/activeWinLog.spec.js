'use strict';

const mockActiveWin = jest.fn();
jest.mock('active-win', () => mockActiveWin);

const { ActiveWinLog } = require('./activeWinLog.js');
const path = require('path');
const userhome = require('userhome');

describe('ActiveWinLog', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('init()', () => {
    it('should set the logFile to the users home directory by default', async () => {
      const awl = new ActiveWinLog({});
      await awl.init();
      expect(awl.logFile).toEqual(path.join(userhome('.active-win-log'), 'awl.json'));
    });
  });

  describe('showHelp()', () => {
    it('should display a help message', () => {
      const awl = new ActiveWinLog({ logFile: 'foo.json' });
      const output = awl.showHelp().split('\n');

      expect(output).toMatchSnapshot();
    });
  });

  describe('getStats()', () => {
    it('should render stats from a (fixture) log file', async () => {
      const awl = new ActiveWinLog({ logFile: path.join(__dirname, '../fixtures/awl.json') });
      const output = await awl.getStats({ lastNDays: 2 });

      expect(output).toMatchSnapshot();
    });
  });

  describe('recordActiveWin()', () => {
    it('should not write to the log file when the active-win library returns a falsey value', async () => {
      const awl = new ActiveWinLog({ logFile: path.join(__dirname, '../fixtures/awl.json') });
      awl.getLogFileData = jest.fn();
      awl.setLogFileData = jest.fn();
      mockActiveWin.mockResolvedValue(null);

      await awl.recordActiveWin();

      // We bailed, so neither method was called
      expect(awl.getLogFileData).not.toHaveBeenCalled();
      expect(awl.setLogFileData).not.toHaveBeenCalled();
    });

    it('should not write to the log file when the active-win library throws an error', async () => {
      const awl = new ActiveWinLog({ logFile: path.join(__dirname, '../fixtures/awl.json') });
      awl.getLogFileData = jest.fn();
      awl.setLogFileData = jest.fn();
      mockActiveWin.mockRejectedValue(new Error('Test: something went wrong'));

      await awl.recordActiveWin();

      // We bailed, so neither method was called
      expect(awl.getLogFileData).not.toHaveBeenCalled();
      expect(awl.setLogFileData).not.toHaveBeenCalled();
    });

    it('should write the active-win window information to the log file', async () => {
      const awl = new ActiveWinLog({ logFile: path.join(__dirname, '../fixtures/awl.json') });
      awl.getLogFileData = jest.fn().mockResolvedValue({ stats: {} });
      awl.setLogFileData = jest.fn();
      mockActiveWin.mockResolvedValue({
        title: 'garbo.org',
        owner: {
          name: 'FireFox',
        },
      });

      await awl.recordActiveWin('2019-03-22T10:45:01.234Z');

      // We bailed, so neither method was called
      expect(awl.getLogFileData).toHaveBeenCalled();
      expect(awl.setLogFileData).toHaveBeenCalledWith({
        stats: {
          '2019-03-22': {
            FireFox: {
              'garbo.org': ['2019-03-22T10:45:01.234Z'],
            },
          },
        },
      });
    });
  });
});
