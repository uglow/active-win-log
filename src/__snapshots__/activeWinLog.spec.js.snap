// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveWinLog getStats() should render stats from a (fixture) log file 1`] = `
"[33m[1m[4m2019-03-20                                          [24m[22m[39m[33m[1m[4m00:00:50[24m[22m[39m
[32mWebStorm                                            [39m[32m[1m00:00:05[22m[39m
  [36mword-domination-plan-703                          [39m[36m00:00:05[39m
[32mSublime                                             [39m[32m[1m00:00:05[22m[39m
  [36mmanifesto.txt                                     [39m[36m00:00:05[39m
[32mGoogle                                              [39m[32m[1m00:00:40[22m[39m
  [36mHow to be good at things                          [39m[36m00:00:40[39m

[33m[1m[4m2019-03-22                                          [24m[22m[39m[33m[1m[4m00:00:40[24m[22m[39m
[32mWebStorm                                            [39m[32m[1m00:00:05[22m[39m
  [36mactive-win-log [/mydev/code/npm/active-win-log]   [39m[36m00:00:05[39m
[32mSublime Text                                        [39m[32m[1m00:00:05[22m[39m
  [36mawl.json                                          [39m[36m00:00:05[39m
[32mGoogle Chrome                                       [39m[32m[1m00:00:30[22m[39m
  [36mtable - npm                                       [39m[36m00:00:05[39m
  [36mCreating a real-world CLI app with Node           [39m[36m00:00:20[39m
  [36m<blank>                                           [39m[36m00:00:05[39m
"
`;

exports[`ActiveWinLog showHelp() should display a help message 1`] = `
Array [
  "",
  "Usage: awl <options>",
  "awl           Start monitoring the active window",
  "awl -?        This help information",
  "awl -l n      Display statistics for the last \\"n\\" days in the log",
  "awl -q        Stop monitoring",
  "Global log location: foo.json",
]
`;
